Routes entity page, router expressions, proxying (#168)
Kong/developer.konghq.com: app/gateway/traffic-control/proxy.md

| `X-Forwarded-Port: <port>`              | `<port>` is the port of the server which accepted a request. If `$realip_remote_addr` is one of the **trusted** addresses, the request header with the same name gets forwarded if provided. Otherwise, the value of the `$server_port` variable provided by [ngx_http_core_module](https://nginx.org/docs/http/ngx_http_core_module.html#var_server_port) will be used. |
| `X-Forwarded-Prefix: <path>`            | `<path>` is the path of the request which was accepted by {{site.base_gateway}}. If `$realip_remote_addr` is one of the **trusted** addresses, the request header with the same name gets forwarded if provided. Otherwise, the value of the `$request_uri` variable (with the query string stripped) provided by [ngx_http_core_module](https://nginx.org/docs/http/ngx_http_core_module.html#var_server_port) will be used. **Note**: {{site.base_gateway}} returns `"/"` for an empty path, but it doesn't do any other normalization on the request path. |
| All other headers | Forwarded as-is by {{site.base_gateway}} | 

> <!--vale on-->

Reviewer: ```suggestion

<!--vale on-->
```

@cloudjumpercat for some reason adding a new line here fixes the rendering issue 🤷

---

> title: Proxying with {{site.base_gateway}}

Author: For reviewers: Is this too much info for the proxying page? Should proxying just contain info about proxying and the connective parts (ex. listeners, routing, plugins, etc) should go on a traffic control landing page with proxying as another card/section instead? This is similar to what we had in the offsite proxy branch: https://github.com/Kong/developer.konghq.com/compare/main...proxy-reference#diff-8922116eb8441698da97efe2beeb185fb024118f8add77981f8eeb930867bbc6

I thought listeners, etc. could be helpful on this page to understanding proxying, but wasn't sure if others felt it was out of place.

description: "Proxying is when {{site.base_gateway}} matches an HTTP request with a [registered route](/gateway/entities/route/) and forwards the request."

content_type: reference
layout: reference

products:
  - gateway

related_resources:
  - text: Route entity
    url: /gateway/entities/route/
  - text: Router Expressions language
    url: /gateway/routing/expressions/
  - text: Expressions repository
    url: https://github.com/Kong/atc-router

breadcrumbs:
  - /gateway/
---

{{ page.description }} This page details information about how {{site.base_gateway}} handles proxying. The following diagram shows how proxying is handled by {{site.base_gateway}}:

> {% mermaid %}

Author: For reviewers: I used ChatGPT help on this diagram, but I revised everything and the input I provided it was the ordered list below this diagram. Still, I'd like a review for tech accuracy, especially the parts about the load balancer. I think ChatGPT got it wrong, so I made some changes to that part and wanted to confirm that I did them correctly.

    LoadBalancer->>UpstreamService: Forwards request to selected instance
    UpstreamService->>Gateway: Sends response
    Gateway->>Plugins: Executes plugins in the `header_filter` phase
    Gateway->>Client: Streams response back to client
{% endmermaid %}

{{site.base_gateway}} handles proxying in the following order:

1. {{site.base_gateway}} listens for HTTP traffic on its configured
proxy port(s) (`8000` and `8443` by default) and L4 traffic on explicitly configured
`stream_listen` ports.
1. {{site.base_gateway}} will evaluate any incoming HTTP request or L4 connection against the Routes you have configured and try to find a matching one.
1. If multiple Routes match, the {{site.base_gateway}} router then orders all defined Routes by their priority and uses the highest priority matching Route to handle a request.
1. If a given request matches the rules of a specific route, {{site.base_gateway}} will
run any global, Route, or Service [plugins]() before it proxies the request. They are run in the following order: Route, Service <!--when does the global config get run?--> These configured plugins will run their `access` phase, which you can find more

> information about in the [Plugin development guide][plugin-development-guide].

Author: For reviewers: I don't think we added the autogenerated PDK yet. How do we want to handle links to autogenerated pages?

subsequent plugins added to the route and/or service that implement a hook in
the `header_filter` phase.

## Listeners

From a high-level perspective, {{site.base_gateway}} listens for HTTP traffic on its configured
proxy ports: `8000` and `8443` by default and L4 traffic on explicitly configured
`stream_listen` ports. {{site.base_gateway}} will evaluate any incoming
HTTP request or L4 connection against the routes you have configured and try to find a matching
one.

{{site.base_gateway}} exposes several interfaces which can be tweaked by the following configuration properties:

- `proxy_listen`: Defines a list of addresses/ports on which {{site.base_gateway}} will
  accept **public HTTP (gRPC, WebSocket, etc) traffic** from clients and proxy
  it to your upstream services (`8000` by default).
- `admin_listen`, which also defines a list of addresses and ports, but those
  should be restricted to only be accessed by administrators, as they expose
  Kong's configuration capabilities: the **Admin API** (`8001` by default).
    {:.important}
   > **Important**: If you need to expose the `admin_listen` port to the internet in a production environment,

>    > {% if_version lte:2.8.x %}[secure it with authentication](/gateway/{{include.release}}/admin-api/secure-admin-api/).{% endif_version %}{% if_version gte:3.0.x %}[secure it with authentication](/gateway/{{include.release}}/production/running-kong/secure-admin-api/).{% endif_version %}

Author: For reviewers: I couldn't find the Admin API in the dev site, does anyone have a link to it?

- `stream_listen`, which is similar to `proxy_listen` but for Layer 4 (TCP, TLS)
  generic proxy. This is turned off by default.

> Kong Gateway is a transparent proxy, and it defaults to forwarding the request to your upstream service untouched, with the exception of various headers such as `Connection`, `Date`, and others as required by the HTTP specifications.

Reviewer 2: ```suggestion
{{site.base_gateway}} is a transparent proxy, and it defaults to forwarding the request to your upstream service untouched, with the exception of various headers such as `Connection`, `Date`, and others as required by the HTTP specifications.
```

the next upstream.

There are two configurable elements:

1. The number of retries. This can be configured per Service using the
   `retries` property. See the [Admin API][API] for more details on this.

2. What exactly constitutes an error. Here {{site.base_gateway}} uses the Nginx defaults, which means an error or timeout that occurs while establishing a connection with the server, passing a request to it, or reading the response headers.

The second option is based on Nginx's 
[`proxy_next_upstream`](https://nginx.org/docs/http/ngx_http_proxy_module.html#proxy_next_upstream) directive. This option is not
directly configurable through {{site.base_gateway}}, but can be added using a custom Nginx
configuration. See the [configuration reference][configuration-reference] for
more details.

### Response streaming

{{site.base_gateway}} receives the response from the upstream service and sends it back to the
downstream client in a streaming fashion. At this point, {{site.base_gateway}} executes
subsequent plugins added to the route or service that implement a hook in
the `header_filter` phase.

Once the `header_filter` phase of all registered plugins has been executed, the

> following headers are added by {{site.base_gateway}} and the full set of headers be sent to

Reviewer 2: ```suggestion
following headers are added by {{site.base_gateway}} and the full set of headers is sent to
```

---
title: Proxying with {{site.base_gateway}}

description: "Proxying is when {{site.base_gateway}} matches an HTTP request with a [registered route](/gateway/entities/route/) and forwards the request."

content_type: reference
layout: reference

products:
  - gateway

related_resources:
  - text: Route entity
    url: /gateway/entities/route/
  - text: Router Expressions language
    url: /gateway/routing/expressions/
  - text: Expressions repository
    url: https://github.com/Kong/atc-router

breadcrumbs:
  - /gateway/

>   - /gateway/traffic-control/

Reviewer: It's failing because this page doesn't exist `/gateway/traffic-control/` , I'll improve the error message

Author: Thanks! I fixed it now
